File: apps/web/src/lib/pipeline/scoring.ts
/**
 * Scoring functions — TypeScript port of worker/scoring.py.
 * Used by the Node.js lite pipeline (Vercel-compatible, no Python dep).
 */

interface OnchainSignal {
  tx_count?: number;
  tx_count_baseline?: number;
  unique_wallets?: number;
  unique_wallets_baseline?: number;
  new_wallet_share?: number;
  new_wallet_share_baseline?: number;
  retention_7d?: number;
  retention_7d_baseline?: number;
}

interface DevSignal {
  commits?: number;
  commits_baseline?: number;
  stars_delta?: number;
  stars_delta_baseline?: number;
  new_contributors?: number;
  new_contributors_baseline?: number;
  releases?: number;
  releases_baseline?: number;
}

interface SocialSignal {
  mentions_count?: number;
  mentions_count_baseline?: number;
  unique_authors?: number;
  unique_authors_baseline?: number;
  engagement_score?: number;
  engagement_score_baseline?: number;
  snippets?: Array<{ text: string; classification?: string }>;
}

export interface MergedSignal {
  key: string;
  label: string;
  kind: string;
  first_seen: string;
  onchain: OnchainSignal;
  dev: DevSignal;
  social: SocialSignal;
}

export interface ScoredCandidate {
  signal: MergedSignal;
  features: Record<string, number>;
  momentum: number;
  novelty: number;
  quality: number;
  totalScore: number;
  normalizedScore: number;
}

function zScore(current: number, baseline: number): number {
  if (baseline === 0) return current > 0 ? 2.0 : 0;
  const z = (current - baseline) / Math.max(baseline, 0.001);
  return Math.max(-5, Math.min(5, z)); // clamp [-5, 5]
}

function computeOnchainFeatures(oc: OnchainSignal): Record<string, number> {
  return {
    z_tx_count: zScore(oc.tx_count ?? 0, oc.tx_count_baseline ?? 0),
    z_unique_wallets: zScore(oc.unique_wallets ?? 0, oc.unique_wallets_baseline ?? 0),
    z_new_wallet_share: zScore(oc.new_wallet_share ?? 0, oc.new_wallet_share_baseline ?? 0),
    z_retention: zScore(oc.retention_7d ?? 0, oc.retention_7d_baseline ?? 0),
  };
}

function computeDevFeatures(dev: DevSignal): Record<string, number> {
  return {
    z_commits: zScore(dev.commits ?? 0, dev.commits_baseline ?? 0),
    z_stars_delta: zScore(dev.stars_delta ?? 0, dev.stars_delta_baseline ?? 0),
    z_new_contributors: zScore(dev.new_contributors ?? 0, dev.new_contributors_baseline ?? 0),
    z_releases: zScore(dev.releases ?? 0, dev.releases_baseline ?? 0),
  };
}

function computeSocialFeatures(social: SocialSignal): Record<string, number> {
  return {
    z_mentions_delta: zScore(social.mentions_count ?? 0, social.mentions_count_baseline ?? 0),
    z_unique_authors: zScore(social.unique_authors ?? 0, social.unique_authors_baseline ?? 0),
    z_engagement_delta: zScore(social.engagement_score ?? 0, social.engagement_score_baseline ?? 0),
  };
}

const WEIGHTS: Record<string, number> = {
  z_tx_count: 0.25,
  z_unique_wallets: 0.2,
  z_new_wallet_share: 0.15,
  z_retention: 0.1,
  z_commits: 0.2,
  z_stars_delta: 0.15,
  z_new_contributors: 0.1,
  z_releases: 0.05,
  z_mentions_delta: 0.15,
  z_unique_authors: 0.1,
  z_engagement_delta: 0.1,
};

function computeMomentum(features: Record<string, number>): number {
  let score = 0;
  for (const [key, weight] of Object.entries(WEIGHTS)) {
    score += (features[key] ?? 0) * weight;
  }
  return score;
}

function computeNovelty(firstSeen: string): number {
  const daysOld =
    (Date.now() - new Date(firstSeen).getTime()) / (1000 * 60 * 60 * 24);
  if (daysOld <= 60) {
    const bonus = 1.3 - 0.3 * (daysOld / 60); // linear decay from 1.3 → 1.0
    return bonus;
  }
  return 1.0;
}

function computeQualityPenalty(
  features: Record<string, number>,
  snippets: Array<{ text: string; classification?: string }>
): number {
  let penalty = 1.0;

  // Single-wallet spike detection
  const walletZ = features.z_unique_wallets ?? 0;
  const txZ = features.z_tx_count ?? 0;
  if (txZ > 2 && walletZ < 0.5) {
    penalty *= 0.6; // likely bot/single-wallet spam
  }

  // Hype-only social penalty
  if (snippets.length > 0) {
    const hypeCount = snippets.filter(
      (s) => s.classification === "hype"
    ).length;
    if (hypeCount / snippets.length > 0.8) {
      penalty *= 0.7; // mostly hype, little substance
    }
  }

  return penalty;
}

export function scoreSignals(signals: MergedSignal[]): ScoredCandidate[] {
  const scored: ScoredCandidate[] = signals.map((sig) => {
    const ocFeatures = computeOnchainFeatures(sig.onchain ?? {});
    const devFeatures = computeDevFeatures(sig.dev ?? {});
    const socialFeatures = computeSocialFeatures(sig.social ?? {});
    const features = { ...ocFeatures, ...devFeatures, ...socialFeatures };

    const momentum = computeMomentum(features);
    const noveltyMult = computeNovelty(sig.first_seen);
    const qualityMult = computeQualityPenalty(
      features,
      (sig.social?.snippets as Array<{ text: string; classification?: string }>) ?? []
    );

    const totalScore = (momentum * noveltyMult) * qualityMult;

    return {
      signal: sig,
      features,
      momentum,
      novelty: noveltyMult,
      quality: qualityMult,
      totalScore,
      normalizedScore: 0,
    };
  });

  // Normalize to [0, 1]
  const totals = scored.map((s) => s.totalScore);
  const minScore = Math.min(...totals);
  const maxScore = Math.max(...totals);
  const range = maxScore - minScore || 1;

  for (const s of scored) {
    s.normalizedScore = (s.totalScore - minScore) / range;
  }

  scored.sort((a, b) => b.totalScore - a.totalScore);
  return scored;
}

File: archive/worker/scoring.py
"""Signal scoring: z-score computation, momentum, novelty, and quality scoring."""

import math
from datetime import datetime, timezone
from config import (
    SCORING_WEIGHTS, NOVELTY_BONUS_DAYS, NOVELTY_BONUS_MULTIPLIER,
    QUALITY_PENALTY,
)


def z_score(current: float, baseline: float, epsilon: float = 1e-6) -> float:
    """Compute a simple z-like ratio score: (current - baseline) / max(baseline, epsilon)."""
    if baseline < epsilon:
        return min(current / epsilon, 10.0)  # cap extreme values
    return (current - baseline) / baseline


def compute_onchain_features(signal: dict) -> dict:
    """Compute z-scores for onchain metrics."""
    return {
        "z_tx_count": z_score(signal["tx_count"], signal["tx_count_baseline"]),
        "z_unique_wallets": z_score(signal["unique_wallets"], signal["unique_wallets_baseline"]),
        "z_new_wallet_share": z_score(signal["new_wallet_share"], signal["new_wallet_share_baseline"]),
        "z_retention": z_score(signal["retention_7d"], signal["retention_7d_baseline"]),
    }


def compute_dev_features(signal: dict) -> dict:
    """Compute z-scores for dev metrics."""
    return {
        "z_commits": z_score(signal["commits"], signal["commits_baseline"]),
        "z_stars_delta": z_score(signal["stars_delta"], signal["stars_delta_baseline"]),
        "z_new_contributors": z_score(signal["new_contributors"], signal["new_contributors_baseline"]),
        "z_releases": z_score(signal["releases"], signal["releases_baseline"]),
    }


def compute_social_features(signal: dict) -> dict:
    """Compute z-scores for social metrics."""
    return {
        "z_mentions_delta": z_score(signal["mentions_count"], signal["mentions_count_baseline"]),
        "z_unique_authors": z_score(signal["unique_authors"], signal["unique_authors_baseline"]),
        "z_engagement_delta": z_score(signal["engagement_score"], signal["engagement_score_baseline"]),
    }


def compute_momentum(features: dict) -> float:
    """Weighted sum of all z-score features using SCORING_WEIGHTS."""
    score = 0.0
    for category, weights in SCORING_WEIGHTS.items():
        for key, weight in weights.items():
            if key in features:
                # Clamp z-scores to [-5, 5] to prevent extreme outliers
                val = max(-5.0, min(5.0, features[key]))
                score += val * weight
    return score


def compute_novelty(first_seen: datetime | str) -> float:
    """Novelty bonus if entity was first seen within NOVELTY_BONUS_DAYS."""
    if isinstance(first_seen, str):
        first_seen = datetime.fromisoformat(first_seen.replace("Z", "+00:00"))
    if first_seen.tzinfo is None:
        first_seen = first_seen.replace(tzinfo=timezone.utc)

    age_days = (datetime.now(timezone.utc) - first_seen).days
    if age_days <= NOVELTY_BONUS_DAYS:
        # Linear decay: full bonus at day 0, zero at NOVELTY_BONUS_DAYS
        return NOVELTY_BONUS_MULTIPLIER * (1.0 - age_days / NOVELTY_BONUS_DAYS)
    return 0.0


def compute_quality_penalty(features: dict, social_snippets: list[dict] | None = None) -> float:
    """
    Detect spam/noise patterns and return a penalty multiplier (0.0 to 1.0).
    1.0 = no penalty, lower = more penalty.
    """
    penalty = 1.0

    # Check for single-wallet spike (if new_wallet_share is very high and retention is low)
    nws = features.get("z_new_wallet_share", 0)
    ret = features.get("z_retention", 0)
    if nws > 2.0 and ret < 0.0:
        # High new wallet influx + poor retention = potential airdrop farming
        penalty *= QUALITY_PENALTY["penalty_multiplier"]

    # Check for single-author hype in social
    if social_snippets:
        hype_count = sum(1 for s in social_snippets if s.get("class") == "hype")
        total = len(social_snippets)
        if total > 0 and hype_count / total > QUALITY_PENALTY["single_author_hype_ratio"]:
            penalty *= QUALITY_PENALTY["penalty_multiplier"]

    return penalty


def compute_total_score(momentum: float, novelty: float, quality_penalty: float) -> float:
    """Final composite score."""
    base = momentum + novelty
    return max(0.0, base * quality_penalty)


def normalize_scores(scores: list[float]) -> list[float]:
    """Normalize a list of scores to [0, 1] range."""
    if not scores:
        return []
    min_s = min(scores)
    max_s = max(scores)
    rng = max_s - min_s
    if rng < 1e-8:
        return [0.5] * len(scores)
    return [(s - min_s) / rng for s in scores]
